fairscale/optim/grad_scaler.py

```diff
@@ -38,36 +81,326 @@ def __init__(
-            )
+        if self._enabled:
+            self._per_optimizer_states = defaultdict(_refresh_per_optimizer_state)
+            self.group = process_group
+
+    def scale(self, outputs: Union[torch.Tensor, List[torch.Tensor]]) -> Union[torch.Tensor, abc.Iterable]:
+        """
+        Multiplies ('scales') a tensor or list of tensors by the scale factor.
+
+        Returns scaled outputs.  If this instance of :class:`GradScaler` is not enabled, outputs are returned
+        unmodified.
+
+        Args:
+            outputs (Tensor or iterable of Tensors):  Outputs to scale.
+        """
+        if not self._enabled:
+            return outputs
 
-        self.display_warning = False  # Only warn once
+        # Short-circuit for the common case.
+        if isinstance(outputs, torch.Tensor):
+            assert outputs.is_cuda or outputs.device.type == "xla" or outputs.device.type == "cpu"
+            if self._scale is None:
+                self._lazy_init_scale_growth_tracker(outputs.device)  # type: ignore
+            assert self._scale is not None
+            return outputs * self._scale.to(device=outputs.device, non_blocking=True)
 
+        # Invoke the more complex machinery only if we're treating multiple outputs.
+        stash: List[_GeneralMultiDeviceReplicator] = []  # holds a reference that can be overwritten by apply_scale
+
+        def apply_scale(val: Union[torch.Tensor, abc.Iterable]) -> Union[torch.Tensor, abc.Iterable]:
+            if isinstance(val, torch.Tensor):
+                assert val.is_cuda or val.device.type == "xla" or val.device.type == "cpu"
+                if len(stash) == 0:
+                    if self._scale is None:
+                        self._lazy_init_scale_growth_tracker(val.device)  # type: ignore
+                    assert self._scale is not None
+                    stash.append(_GeneralMultiDeviceReplicator(self._scale))
+                return val * stash[0].get(val.device)
+            elif isinstance(val, abc.Iterable):
+                iterable = map(apply_scale, val)
+                if isinstance(val, list) or isinstance(val, tuple):
+                    return type(val)(iterable)
+                else:
+                    return iterable
+            else:
+                raise ValueError("outputs must be a Tensor or an iterable of Tensors")
+
+        return apply_scale(outputs)
+
+    def _foreach_non_finite_check_and_unscale_cpu_(
+        self, grads: List, found_inf: torch.Tensor, inv_scale: torch.Tensor
+    ) -> None:
+        if len(grads) == 0:
+            return
+        assert inv_scale.numel() == 1, "inv_scale must be a 1-element tensor."
+        assert found_inf.numel() == 1, "found_inf must be a 1-element tensor."
+
+        expected_device = grads[0].device
+        for tensor in grads[0]:
+            assert tensor.device == expected_device, "grads must be on the same device"
```

Comment:
also print the devices of the tensor and expected device for easier debugging.

Reply:
wrapped the assert statement in a try/except block. If the assertion is raised then both the tensor device and expected devices will be logged.